### DISCUSSION TITLE
2.9.0-preview1

### DISCUSSION THREAD

**Discussion by glopesdev** [bonsai-rx/bonsai#1991] [Announcements]
> [!WARNING]
> **THIS RELEASE IS UNDER ACTIVE DEVELOPMENT.** We are actively looking for feedback on your experience with the new features. **All preview features are subject to breaking changes as we incorporate feedback** and so _**we do not recommend you to use this release**_ for any production applications.

This is the first preview release in the 2.9 series, which introduces significant changes to the language editor experience. Please give us any feedback either in the comments below or by opening a new discussion. If you create a new discussion, please indicate clearly that your question is about the 2.9.0-preview1 release.

In this preview, we are introducing [a new workflow navigation model](https://github.com/bonsai-rx/bonsai/pull/1870) coupling a single-page browser-like view of the current workflow level, with a nested hierarchy tree view and breadcrumbs to allow quickly jumping between the different sections and modules of a complex project.

We also introduce [a new workflow state watch mode](https://github.com/bonsai-rx/bonsai/pull/1975), active by default, which provides an unprecedented level of runtime feedback about the inner workings of a reactive workflow, including when sequences are subscribed to, when they produce values, terminate, or are cancelled. We expect this mode to provide an invaluable tool for learning visual reactive programming and understanding the behavior of programs written in Bonsai.

In anticipation of new editor features, we are introducing [a reorganization of the `.layout` files](https://github.com/bonsai-rx/bonsai/pull/1956), which were previously used to store debugger visualizer settings over multiple runs. These files previously cluttered folders containing multiple workflow files and are hard to extend to the more complex project-based settings we now need to evolve the editor experience. With this reorganization we expect to facilitate a cleaner project folder structure and introduce more clarity on the version control of workflows.

Finally, we [modernized the package manager dialog](https://github.com/bonsai-rx/bonsai/pull/1969) to support our migration from MyGet to NuGet. [As discussed during the last MyGet outage](https://github.com/orgs/bonsai-rx/discussions/1500), we urgently need to broaden our community package deployment options. The new package manager introduces support for two new [package types](https://learn.microsoft.com/en-us/nuget/create-packages/set-package-type?tabs=dotnet): `BonsaiLibrary` and `BonsaiGallery` for packages providing libraries of Bonsai operators and example workflows, respectively.

These package types will be used to curate the `nuget.org` feed to keep our package manager focused on libraries providing reactive operators. Currently package type filtering is only supported for the `nuget.org` feed, but it is turned on by default there. New packages deployed to `nuget.org` are expected to declare both `Dependency` and `BonsaiLibrary` as their package types for maximum discoverability.

> [!Important]
> In the near future, we will move entirely out of MyGet and will archive the community feed.

### 🔮 Feature Preview

* Modernize package manager dialog by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1969
* Move layout files to workflow settings folder by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1956
* Add workflow state watch to debug mode by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1975
* Rewrite editor navigation model by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1870

### 🎉 New Features

* Allow propagation of Defer nested sequence visualizer by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1931
* Rename CombineTimestamp to CreateTimestamped by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1941
* Make period a nullable property of reactive timer and clarify documentation by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1957

### 🐞 Bug Fixes

* Exclude build dependencies when counting inputs to new nested node by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1946

### 🚀 Performance Improvements

* Replace uses of WebRequest with HttpClient by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1937

### 🔧 Engineering

* Update launch settings to target artifacts folder by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1923
* Update ScintillaNET to support modern .NET by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1924
* Allow building workflows as typed sequences by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1942
* Target netcore on editor and bootstrapper projects by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1907
* Configure tests to run on .NET 8 by @PathogenDavid in https://github.com/bonsai-rx/bonsai/pull/1952
* Fix test target framework issues by @PathogenDavid in https://github.com/bonsai-rx/bonsai/pull/1958
* Use deployment key in CI for manipulating Git repository by @PathogenDavid in https://github.com/bonsai-rx/bonsai/pull/1985
* Add support for sorting test workflows by @glopesdev in https://github.com/bonsai-rx/bonsai/pull/1964

**Full Changelog**: https://github.com/bonsai-rx/bonsai/compare/2.8.5...2.9.0-preview1

<hr /><em>This discussion was created from the release <a href='https://github.com/bonsai-rx/bonsai/releases/tag/2.9.0-preview1'>2.9.0-preview1</a>.</em>

**Comment by jsiegle** [comment]
Thanks for all your hard work on this, the Explorer View and State Watch features turned out great!

I have two bits of feedback on the Explorer:

1. Breadcrumb icons

Not sure if this is intentional or not, but the breadcrumb icons are asymmetrical, making them look cut off at the bottom:

![explorer_tabs](https://github.com/user-attachments/assets/13897af9-6131-47b7-8d1a-ec9799701fbf)

I think it would look better if the border went all the way around.

2. Opening new workflow windows

In some cases (especially with the new state watch feature) it is helpful to have multiple parts of the workflow open side-by-side in separate windows. Would it be possible to add back the ability to open new windows (perhaps via the menu that's opened after right-clicking on the group workflow)? If there are multiple windows open, navigating via the explorer would only apply to the main window.

**Comment by glopesdev** [comment]
@jsiegle Thanks for the feedback!

> Not sure if this is intentional or not, but the breadcrumb icons are asymmetrical, making them look cut off at the bottom:

Oops, definitely not intentional and likely an oversight in scaling tests. I have opened an issue for this (https://github.com/bonsai-rx/bonsai/issues/2000) and will correct in the next preview.

> Would it be possible to add back the ability to open new windows (perhaps via the menu that's opened after right-clicking on the group workflow)? If there are multiple windows open, navigating via the explorer would only apply to the main window.

I added the feedback directly as a comment on the navigation rewrite PR (https://github.com/bonsai-rx/bonsai/pull/1870#issuecomment-2331596842). This is something we definitely want to have and want to iterate on for the following previews.

**Comment by jonnew** [comment]
With these new breadcrumb feature, is there any use of the space next to the main tab?

![image](https://github.com/user-attachments/assets/6a10fd01-4452-4731-b36d-b74a39b63d10)

I find this pretty confusing to use to be honest. It feels like it's making the top level Tab irrelevant and the word "Workflow" is kinda getting repeated all over the place. I think maybe the explorer is all thats needed along with modifying the top level tab name with the heirarchy:

![image](https://github.com/user-attachments/assets/9915c675-d16b-4527-aab8-3befe2ddd94a)

**Reply by glopesdev** [reply]
@jonnew Thanks for the feedback, the space is kept since our goal is to allow "pinning" or opening multiple views into different parts of the project tree simultaneously, hopefully by the time the stable 2.9 is out (see https://github.com/bonsai-rx/bonsai/pull/1870#issuecomment-2331596842).

Ideally we would make this dockable so it is possible to switch between floating and docked windows at will, but we'll have to see how far we get with that.

As for the breadcrumbs, they are named differently from the moment you either save your workflow, or load an existing one, with the start breadcrumb showing the name of the project. However, here we are considering changing this to a "Home" icon instead to save space. We will be experimenting with this for the next month or so, any ideas welcome.

**Comment by cjsha** [comment]
This button indicates package updates are available even when there aren't actually any package updates available. 

![Screenshot 2024-09-06 110523](https://github.com/user-attachments/assets/7c2624bd-aa13-47e8-b7c6-9413094f9721)

I suppose it should be greyed out with a different tooltip i.e. "No package updates available" when there are no package updates available?

I like the idea behind this feature a lot btw.

**Reply by glopesdev** [reply]
@cjsha Thanks for the feedback, I have reopened your issue https://github.com/bonsai-rx/bonsai/issues/2003. It's fine to open issues as well, we can move them back and forth from discussions as needed.

**Comment by cjsha** [comment]
While you're making adjacent changes for the Bonsai 2.9 release, I propose adding a "Show Deprecated" checkbox that is disabled by default.

For example:
![Screenshot 2024-09-06 104044](https://github.com/user-attachments/assets/38aa50c1-6fd8-4544-b0e8-ad08d6cd229f)

**Reply by glopesdev** [reply]
@cjsha What would be the goal of showing only deprecated packages? Drawing attention to deprecated packages seems like the last thing we would want to do in the search, both from the side of producers and consumers.

In any case, the nuget.org [search API](https://learn.microsoft.com/en-us/nuget/api/search-query-service-resource) does not support querying for deprecated packages, so this will not be possible to implement in practice.

**Reply by cjsha** [reply]
The goal would be to hide deprecated packages by default, not show only deprecated packages.

**Reply by glopesdev** [reply]
Ah, I see now. Unfortunately the search API does not allow for this.

The recommended way to do this on NuGet is to "unlist" the package from search entirely.

For now since all packages will not have package types set all currently deprecated packages will be hidden by default.
